ENH: Add nd-support to trim_zeros

lagru opened the pull request:
Add support for trimming nd-arrays with trim_zeros while preserving the old behavior for 1D input. The new parameter `axis` can specify a single dimension to be trimmed (reducing all other dimensions to the envelope of absolute values). If None or multiple values are specified, all or the selected dimensions are trimmed iteratively.

Additionally provide the `atol`, `rtol` and `return_lengths` parameters. The first two control what is considered  a "zero" to be trimmed, the latter provides the user with the on how much was trimmed.

I recently needed this behavior myself and think this makes the function applicable to more use cases.

This is currently marked as work in progress, as I wasn't sure about the new API. I tried to preserve the old API but I think it's worth having a discussion about refactoring and additional API changes. See also the review-comments below.

lagru committed:
ENH: Add nd-support to trim_zeros

Add support for trimming nd-arrays with trim_zeros while preserving the
old behavior for 1D input. The new parameter `axis` can specify a single
dimension to be trimmed (reducing all other dimensions to the envelope
of absolute values). If None or multiple values are specified, all or
the selected dimensions are trimmed iteratively.

This should make the function applicable to more use cases.

Additionally provide the `atol`, `rtol` and `return_lengths` parameters.
The first two control what is considered  a "zero" to be trimmed, the
latter provides the user with the on how much was trimmed.

lagru commented on numpy/lib/function_base.py:
```diff
@@ -1557,28 +1557,55 @@ def sort_complex(a):
+def trim_zeros(
+    filt,
+    trim='fb',
+    axis=-1,
```
Not sure about the proper default value here. `None` (trim all dimensions), 0 or 1 (trim first/last dimension) are candidates that would preserve the old behavior.

lagru commented on numpy/lib/function_base.py:
```diff
@@ -1557,28 +1557,55 @@ def sort_complex(a):
+        Relative tolerance with which a value is considered for trimming.
+    return_lengths : bool, optional
+        Additionally return the number of trimmed samples in each dimension at
+        the front and back.
```
I don't like functions that modify their number of returned values with parameters. It seems like a good idea to provide the user with the trim lengths but in the long run I'd like to see a depreciation of this parameter and that `trim_zeros` always returns trim lengths.

eric-wieser replied:
It's a lousy name, but `arg_trim_zeros` might be a better name for the function returning indices.

lagru replied:
Not enthusiastic about both options. Is there a way to deprecate the number of returned values? E.g. start with default value `False` and warn when `return_lengths=False` is given. Then after a deprecation cycle that parameter could be removed. If you don't think its worth going through a deprecation for this I'd prefer keeping this parameter over adding another utility function.

In case we'd add a separate function `arg_trim_zeros` I would agree that returning the actual indices would be more useful. However I think the current behavior is more fitting because I think the primary use case of returning the indices (slicing) is already taken care of inside the function. I imagine this argument to be useful for cases where a user is interested in the size of the shared zero area. E.g. my original reason for implementing this was to reduce the size of a bunch of nd-signals while still being able to reconstruct the original signal by padding. I see this as an analogue to the `pad_width` parameter of `numpy.pad`.

eric-wieser replied:
> because I think the primary use case of returning the indices (slicing) is already taken care of inside the function

What about the case where I want to trim one array based on the non-zero area of another?

lagru commented on numpy/lib/function_base.py:
```diff
@@ -1595,22 +1622,62 @@ def trim_zeros(filt, trim='fb'):
+
+        # Find start and stop indices for current dimension
+        start, stop = np.nonzero(reduced)[0][[0, -1]]
+        stop += 1
```
These two lines replace the old behavior that would search for `start` and `stop` with two simple for loops. I haven't benchmarked it yet but will do so. 

However, as this is likely the bottleneck of the function, I'd like to do this (as well as the reduction to the absolute envelope above) in Cython (or C). Would that be appreciated or is the lack of compiled code in the lib module deliberate?

lagru committed:
MAINT: Address empty and all-zero input

in trim_zeros.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1624,46 +1624,45 @@ def trim_zeros(
+            start = stop = absolutes.shape[current_axis]
+            if "f" not in trim:
+                # except when only the backside is to be sliced
+                stop = 0
```
I don't really understand what you mean here - are you saying to slice _leaving_ only the front, or slice _removing_ the front?

When asked to slice both ends of an all zero array, I think I'd expect either:
* Only the back to be sliced
* The front to be sliced if `'fb'` is passed, and the back to be sliced if `'bf'` is passed

lagru replied:
Good call. I prefer the latter behavior and have modified the code accordingly. 

The old function was really liberal with what it accepted as input for the `trim` parameter (e.g. only checking `"f" in trim`). I think restricting what is allowed to be passed in would make that parameter easier to understand and document.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1624,46 +1624,45 @@ def trim_zeros(
+    nonzero = np.nonzero(absolutes)
 
     for current_axis in axis:
         absolutes.take([], current_axis)  # Raises if axis is out of bounds
```
Better to just call `axis = normalize_axis_index(axis, absolute.ndim)` here

lagru replied:
Nice, that function is indeed exactly what I needed.

lagru committed:
MAINT: Honor trim order

In case the user passes in an all-zero array and string to `trim` that
starts with "b", the last dimension should be sliced first. In all other
cases the first dimension takes precedence.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1557,28 +1557,59 @@ def sort_complex(a):
+    atol : float, optional
+        Absolute tolerance with which a value is considered for trimming.
+    rtol : float, optional
+        Relative tolerance with which a value is considered for trimming.
```
I don't really like these arguments here - it seems like you're after a function for replacing almost-zeros with zeros, which can then be composed, as `trim_zeros(zero_small_values(x, rtol=...), axis=1)` or similar.

lagru replied:
I don't mind removing them. I added them for convenience but it's easy to assign zeros appropriately before calling this function. 

Seeing that there is already [numpy.isclose](https://docs.scipy.org/doc/numpy/reference/generated/numpy.isclose.html) I'm not convinced that a another function is needed.

lagru committed:
MAINT: Remove atol and rtol from trim_zeros

It's easy to emulate this behavior by assigning zeros appropriately
beforehand.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1595,22 +1606,58 @@ def trim_zeros(filt, trim='fb'):
+    # Normalize axes to 1D-array
+    axis = np.asarray(axis, dtype=np.intp)
+    if axis.ndim == 0:
+        axis = np.asarray([axis], dtype=np.intp)
```
I think you want `normalize_axis_tuple` here, which also removes the need to call `normalize_axis_index` in the loop.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1595,22 +1606,58 @@ def trim_zeros(filt, trim='fb'):
-                break
+    trim = trim.lower()
+
+    absolutes = np.abs(filt)
```
No point taking the absolute value here

eric-wieser commented:
An alternate spelling with vectorization:
```python
nonzero = np.argwhere(filt)
if len(nonzero) == 0:
    if trim.startswith("b"):
        start = stop = np.zeros(filt.ndim)
    else:
        start = stop = np.array(filt.shape)
else:
    start = nonzero.min(axis=0)
    stop = nonzero.max(axis=0)

# TODO: make a function that just returns `start` and `stop`?

sl = tuple(
    slice(a, b) if ax in axis else slice(None)
    for ax, (a, b) in enumerate(zip(start, stop))
)
return filt[sl]

```

lagru committed:
ENH: Add arg_trim_zeros

as it's own function. trim_zeros uses its output to newly support the
nd-case.

lagru committed:
TST: Test basic nd-support for trim_zeros

lagru commented:
@eric-wieser  Concerning the discussion in https://github.com/numpy/numpy/pull/15181#discussion_r361637225 and your other suggestions: You seem to be in favor of adding a new function (e.g. `arg_trim_zeros`) that returns indices to the zero area. I've pushed a draft so that we can discuss that API...

lagru committed:
ENH: Support trimming nd-arrays that are all-zero

lagru committed:
MAINT: Address attribute error in trim_zeros

lagru committed:
MAINT: Make arg_trim_zeros output consistent

Ensure that the returned start and stop arrays of indices have exactly
one dimension.

lagru commented:
@eric-wieser what do you think of the currently proposed API? I'll add missing tests if you're happy with it and give a notice (required I think?) on the mailing list.

An alternative to `arg_trim_zeros` would be to remove it but document its implementation as a lightweight example inside `trim_zeros`'s Note section. It basically amounts to "use `nonzero` and call `max`, `min` on the appropriate dimension of the output".

mattip commented:
We discussed this at a triage meeting, and see no reason to reject it.
@lagru if you are interesting in moving this forward, could you rebase it? Maybe coming to a meeting would also help us move this up the priority queue.

lagru commented:
Thanks for the ping. I'm happy to pick this up again, though, not before next week. Once this is up to speed, I'm happy to join a meeting if that would help.

lagru committed:
Merge branch 'main' into trim_zeros

lagru committed:
Update trim_zeros annotation in stub file

lagru committed:
Make `_arg_trim_zeros` private and add tests

lagru committed:
Trim on all dimensions by default

lagru committed:
Make sure `None` is treated as non-zero

For some reason `argwhere` treats None as zero which fails an existing
test.

lagru commented on numpy/lib/_function_base_impl.py:
```diff
@@ -1867,62 +1867,130 @@ def sort_complex(a):
 
 
-def _trim_zeros(filt, trim=None):
+def _arg_trim_zeros(filt):
```
Made this one private for now. Not sure if this function is valuable enough to increase the API surface.

lagru commented on numpy/lib/_function_base_impl.py:
```diff
@@ -1867,62 +1867,130 @@ def sort_complex(a):
+    >>> _arg_trim_zeros(np.array([0, 0, 1, 1, 0]))
+    (array([2]), array([3]))
+    """
+    nonzero = np.argwhere(filt != 0)
```
Had to use `filt != 0` here instead of just passing `filt` because `np.argwhere` (and `nonzero`) treat `None` as zero. This behavior is also covered by an already existing test [`test_no_trim`](https://github.com/numpy/numpy/blob/cecc005ff2b2ac133f0f535fa86c781dfcb9d1ae/numpy/lib/tests/test_function_base.py#L1391-L1394).

I'm a bit surprised by this behavior. Not sure if there's a faster way than adding the extra `filt != 0` check.

lagru replied:
Only do this if it's an object array.

lagru commented on numpy/lib/_function_base_impl.pyi:
```diff
@@ -316,6 +316,7 @@ def sort_complex(a: ArrayLike) -> NDArray[complexfloating[Any, Any]]: ...
 def trim_zeros(
     filt: _TrimZerosSequence[_T],
     trim: L["f", "b", "fb", "bf"] = ...,
```
Do we want to keep supporting `trim=""`? It's not documented or tested for but worked with the previous implementation...

lagru committed:
Remove test for unsupported `trim=""` for now

lagru commented:
Alright, I rebased this.

> Maybe coming to a meeting would also help us move this up the priority queue.

Happy to attend a meeting, but I'm not sure which meeting type would be appropriate. I could make the next triage call on Sep 18 at 18:00 UTC?

mattip commented:
> I could make the next triage call 

Great!

lagru commented on numpy/lib/_function_base_impl.py:
```diff
@@ -1867,62 +1867,130 @@ def sort_complex(a):
-        array.
+        back. By default, zeros are trimmed from the front and back.
+    axis : int or sequence, optional
+        The axis to trim. If None, the default, all axes are trimmed.
```
Should clarify that this will not produce a ragged array.

lagru replied:
The docstring now explicitly mentions that the return value will be an array with the same number of dimensions and the same dtype. Let me know if this isn't clear enough.

lagru commented on numpy/lib/_function_base_impl.py:
```diff
@@ -1867,62 +1867,130 @@ def sort_complex(a):
         A string with 'f' representing trim from front and 'b' to trim from
-        back. Default is 'fb', trim zeros from both front and back of the
-        array.
+        back. By default, zeros are trimmed from the front and back.
```
Explain what "front" and "back" mean in higher dimensions.

lagru commented on numpy/lib/_function_base_impl.py:
```diff
@@ -1867,62 +1867,130 @@ def sort_complex(a):
 
+    Notes
+    -----
+    For all-zero arrays, the first axis is trimmed first.
```
Also note how this function can be applied for values other than zero.

lagru replied:
After playing around with masked arrays. Even if I do something about the internal call to `np.asarray` that converts the masked array back

```python
import numpy as np
b = np.array([[1, 1, 2, 3, 1, 1],
              [1, 1, 0, 3, 0, 1],
              [1, 1, 1, 1, 1, 1]])
b_masked = np.ma.MaskedArray(b, mask=(b == 1))
np.trim_zeros(b_masked)
```
will still trim zeros as well, in this case the zero at `b[1, 4]`. It shouldn't in a case where we only want to trim ones.

The alternative I see is to make the currently internal function `_arg_trim_zeros` public. But might be another PR.

seberg replied:
I am not sure where the masked array discussion comes from?  I can see a `mask=` being passed in, in principle (to be used instead of `values !=/== 0`), but we shouldn't worry about masked arrays unless it is to not break what currently happens to work.

lagru replied:
That was a concern raised during the triage meeting. The question was along the lines of "Why is zero special? Why shouldn't this work for other values?". 

The suggestion was to explore solutions with masked arrays and document it as an example.

ngoldbaum replied:
Maybe just add a note about the behavior with masked arrays in the docstring?

lagru committed:
Clarify ND in trim_zeros' docstring

Addresses feedback during the triage meeting.

lagru committed:
Optimize cases where trim_zero isn't given an object array

mattip commented:
Thanks @lagru. We can iterated on this if there are problems since it is extending the API.

mattip commented:
ohh, this should have a release note. @lagru could you add one?

lagru commented:
@mattip Yes, I'll add a release note. :)
